src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/EventHandler.java

```diff
@@ -96,6 +100,32 @@ public void run() {
     private boolean handleEvent(Event event) {
         notifier.receivedEvent(event);
 
+        /*
+         * See if the event thread is a vthread that we need to start tracking.
+         */
+        ThreadReference eventThread = null;
+        if (event instanceof ClassPrepareEvent) {
+            eventThread = ((ClassPrepareEvent)event).thread();
+        } else if (event instanceof LocatableEvent) {
+            eventThread = ((LocatableEvent)event).thread();
+        }
+        if (eventThread != null) {
+            // This is a vthread we haven't seen before, so add it to our list.
+            assert eventThread.isVirtual();
```

Comment:
It is not clear why a virtual thread is expected here.
At least, some comment would be nice to have.

Reply:
I looked at the jdb changes. They look good to me. I've posed one comment though.

Reply:
Yeah, that got misplaced, probably when I shuffled around some code. I think it would be correct if it was moved below the `addThread()` call and was changed to:

`            assert !(added && !eventThread.isVirtual());`

which is probably more easily read as:

`            assert !added || eventThread.isVirtual());`

First I need to see if I can even get asserts enabled since that one should have been triggered.

Reply:
Actually that's not quite right either. It looks like it needs to be:

```
            if (added) {
                assert eventThread.isVirtual();
            }
```
If this thread was added, it must be an a virtual thread, because if it was not it would have been previously added by a ThreadStart event.

Reply:
It is better to add a comment explaining this.
Otherwise, it is hard to understand why this conclusion was made.
Also, why do you think this is not equal of your conditional assert above?:
   assert !added || eventThread.isVirtual());

Reply:
> It is better to add a comment explaining this.
> Otherwise, it is hard to understand why this conclusion was made.

I added some comments.

> Also, why do you think this is not equal to your conditional assert above?:
> assert !added || eventThread.isVirtual());

Actually that does work, but I think it is easier to read with the "if" statement.